chacha20/src/backends/neon.rs

```diff
@@ -354,3 +384,259 @@ impl<R: Unsigned> StreamBackend for Backend<R> {
+
+                r0_1 = rotate_left!(r0_1, 7);
+                r1_1 = rotate_left!(r1_1, 7);
+                r2_1 = rotate_left!(r2_1, 7);
+                r3_1 = rotate_left!(r3_1, 7);
+
+                r0_1 = extract!(r0_1, 3);
+                r0_2 = extract!(r0_2, 2);
+                r0_3 = extract!(r0_3, 1);
+
+                r1_1 = extract!(r1_1, 3);
+                r1_2 = extract!(r1_2, 2);
+                r1_3 = extract!(r1_3, 1);
+
+                r2_1 = extract!(r2_1, 3);
+                r2_2 = extract!(r2_2, 2);
+                r2_3 = extract!(r2_3, 1);
+
+                r3_1 = extract!(r3_1, 3);
+                r3_2 = extract!(r3_2, 2);
+                r3_3 = extract!(r3_3, 1);
+            }
+
+            r0_0 = vaddq_u32(r0_0, self.state[0]);
+            r0_1 = vaddq_u32(r0_1, self.state[1]);
+            r0_2 = vaddq_u32(r0_2, self.state[2]);
+            r0_3 = vaddq_u32(r0_3, self.state[3]);
+
+            r1_0 = vaddq_u32(r1_0, self.state[0]);
+            r1_1 = vaddq_u32(r1_1, self.state[1]);
+            r1_2 = vaddq_u32(r1_2, self.state[2]);
+            r1_3 = vaddq_u32(r1_3, self.state[3]);
+            r1_3 = add64!(r1_3, ctrs[0]);
+
+            r2_0 = vaddq_u32(r2_0, self.state[0]);
+            r2_1 = vaddq_u32(r2_1, self.state[1]);
+            r2_2 = vaddq_u32(r2_2, self.state[2]);
+            r2_3 = vaddq_u32(r2_3, self.state[3]);
+            r2_3 = add64!(r2_3, ctrs[1]);
+
+            r3_0 = vaddq_u32(r3_0, self.state[0]);
+            r3_1 = vaddq_u32(r3_1, self.state[1]);
+            r3_2 = vaddq_u32(r3_2, self.state[2]);
+            r3_3 = vaddq_u32(r3_3, self.state[3]);
+            r3_3 = add64!(r3_3, ctrs[2]);
+
+            vst1q_u8(blocks.offset(0), vreinterpretq_u8_u32(r0_0));
+            vst1q_u8(blocks.offset(16), vreinterpretq_u8_u32(r0_1));
+            vst1q_u8(blocks.offset(2 * 16), vreinterpretq_u8_u32(r0_2));
+            vst1q_u8(blocks.offset(3 * 16), vreinterpretq_u8_u32(r0_3));
+
+            vst1q_u8(blocks.offset(64), vreinterpretq_u8_u32(r1_0));
+            vst1q_u8(blocks.offset(16 + 64), vreinterpretq_u8_u32(r1_1));
+            vst1q_u8(blocks.offset(2 * 16 + 64), vreinterpretq_u8_u32(r1_2));
+            vst1q_u8(blocks.offset(3 * 16 + 64), vreinterpretq_u8_u32(r1_3));
+
+            vst1q_u8(blocks.offset(128), vreinterpretq_u8_u32(r2_0));
+            vst1q_u8(blocks.offset(16 + 128), vreinterpretq_u8_u32(r2_1));
+            vst1q_u8(blocks.offset(2 * 16 + 128), vreinterpretq_u8_u32(r2_2));
+            vst1q_u8(blocks.offset(3 * 16 + 128), vreinterpretq_u8_u32(r2_3));
+
+            vst1q_u8(blocks.offset(192), vreinterpretq_u8_u32(r3_0));
+            vst1q_u8(blocks.offset(16 + 192), vreinterpretq_u8_u32(r3_1));
+            vst1q_u8(blocks.offset(2 * 16 + 192), vreinterpretq_u8_u32(r3_2));
+            vst1q_u8(blocks.offset(3 * 16 + 192), vreinterpretq_u8_u32(r3_3));
+
+            self.state[3] = add64!(self.state[3], ctrs[3]);
+        }
```

Comment:
The duplication here is probably the last major remaining concern.

Reply:
Should be a bit less duplication there now

Reply:
I wasn't quite sure how you wanted the duplication reduced... since there are 3 ways that I'm aware of:
1) restructure the code into functions, so that the duplicate code mainly consists of function calls
2) use one method that takes `dest_ptr: *mut u8, num_blocks: usize` as parameters
3) or both

I just chose (1), but I can switch it to either (2) or (3) if you would prefer.